### PR TITLE
New bit ternary

### PR DESCRIPTION
**Pull request by jtlap**

### CODE REVIEW

File: include/eve/module/core/regular/bit_ternary.hpp [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+#include <eve/module/core/regular/bit_xor.hpp>
+#include <iostream>
+
+namespace eve
+{
+  template<typename Options>
+  struct bit_ternary_t : callable<bit_ternary_t, Options>
+  {
+    template<int K, value T0, value T1, value T2>
+    requires(eve::same_lanes_or_scalar<T0, T1, T2> && 0 <= K && K <= 255)
```
Comment (jfalcou) [inline include/eve/module/core/regular/bit_ternary.hpp:27]: Why not taking K as an std::uint8t directly ?

File: include/eve/module/core/regular/bit_ternary.hpp [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+//!    **Return value**
+//!
+//!      1. Bitwise ternary logic that provides the capability to implement any three-operand binary function;
+//!        the specific binary function is specified by value of k.
+//!        For each bit in each integer, the corresponding bit from a, b, and c are used according
+//!        to ik, and the result is output.
+//!      2. equivalent to the call on the elements of the tuple.
+//!      3. [The operation is performed conditionnaly](@ref conditional).
+//!
+//!      This is the pattern of a truth table:
```
Comment (jfalcou) [inline include/eve/module/core/regular/bit_ternary.hpp:81]: This piece of the doc is a bit unclear and need work

File: include/eve/module/core/regular/bit_ternary.hpp [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    // this is adapted from Samuel neves ternary logic for sse avx etc.
+    template < int K, typename T0, typename T1, typename T2, callable_options O>
+    EVE_FORCEINLINE  /*bit_value_t<T0, T1, T2>*/ auto bit_ternary_(EVE_REQUIRES(cpu_)
+                                                                  , O const &
+                                                                  , std::integral_constant<int, K> const &
+                                                                  , [[maybe_unused]] T0 const & aa
+                                                                  , [[maybe_unused]] T1 const & bb
+                                                                  , [[maybe_unused]] T2 const & cc
+                                                                  ) noexcept
+    //    requires(K <= 255)
```
Comment (jfalcou) [inline include/eve/module/core/regular/bit_ternary.hpp:120]: remove allt he comment here and at line 113

File: include/eve/module/core/regular/bit_ternary.hpp [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+  inline constexpr auto bit_ternary = functor<bit_ternary_t>;
+//================================================================================================
+//! @}
+//================================================================================================
+
+  namespace detail
+  {
+
+    // this is adapted from Samuel neves ternary logic for sse avx etc.
+    template < int K, typename T0, typename T1, typename T2, callable_options O>
```
Comment (jfalcou) [inline include/eve/module/core/regular/bit_ternary.hpp:112]: K shoudn be std::uitn8_t

File: test/doc/core/bit_ternary.cpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
+// revision 0
+#include <eve/module/core.hpp>
+#include <iostream>
+
+template < int a, int b, int c, int d, int e, int f , int g, int h>
```
Comment (jfalcou) [inline test/doc/core/bit_ternary.cpp:5]: Yes, this should be public API and need a proper name and API

File: include/eve/module/core/regular/impl/simd/x86/bit_ternary.hpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+#include <eve/detail/category.hpp>
+#include <eve/forward.hpp>
+#include <eve/module/core/constant/mzero.hpp>
+#include <eve/module/core/constant/zero.hpp>
+#include <eve/module/core/regular/bit_notand.hpp>
+#include <eve/module/core/regular/if_else.hpp>
+
+namespace eve::detail
+{
+  template<int K, arithmetic_scalar_value T, typename N, callable_options O>
```
Comment (jfalcou) [inline include/eve/module/core/regular/impl/simd/x86/bit_ternary.hpp:22]: `std::uint8_t`

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: include/eve/module/core/regular/bit_ternary.hpp (hunk 1 above)
Comment (jtlap) [reply include/eve/module/core/regular/bit_ternary.hpp:27]: done

File: include/eve/module/core/regular/bit_ternary.hpp (hunk 2 above)
Comment (jtlap) [reply include/eve/module/core/regular/bit_ternary.hpp:81]: done

File: include/eve/module/core/regular/bit_ternary.hpp (hunk 3 above)
Comment (jtlap) [reply include/eve/module/core/regular/bit_ternary.hpp:120]: done

File: include/eve/module/core/regular/bit_ternary.hpp (hunk 4 above)
Comment (jtlap) [reply include/eve/module/core/regular/bit_ternary.hpp:112]: done

File: test/doc/core/bit_ternary.cpp (hunk 5 above)
Comment (jtlap) [reply test/doc/core/bit_ternary.cpp:5]: done eve;;truth_table

File: include/eve/module/core/regular/impl/simd/x86/bit_ternary.hpp (hunk 6 above)
Comment (jtlap) [reply include/eve/module/core/regular/impl/simd/x86/bit_ternary.hpp:22]: done

File: include/eve/module/core/regular/bit_ternary.hpp [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+#include <eve/module/core/regular/bit_xor.hpp>
+#include <iostream>
+
+namespace eve
+{
+  template<typename Options>
+  struct bit_ternary_t : callable<bit_ternary_t, Options>
+  {
+    template<std::uint8_t K, value T0, value T1, value T2>
+    requires(eve::same_lanes_or_scalar<T0, T1, T2> && 0 <= K && K <= 255)
```
Comment (jfalcou) [inline include/eve/module/core/regular/bit_ternary.hpp:27]: Remove the requires on size of K.

**Review by jfalcou — CHANGES_REQUESTED** [review]

**Review by jfalcou — APPROVED** [review]

**Pull request merged**